ipld/hamt/src/hamt.rs

```diff
@@ -201,4 +201,35 @@ where
     pub fn is_empty(&self) -> bool {
         self.root.is_empty()
     }
+
+    /// Iterates over each KV in the Hamt and runs a function on the values.
+    ///
+    /// This function will constrain all values to be of the same type
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use ipld_hamt::Hamt;
+    ///
+    /// let store = db::MemoryDB::default();
+    ///
+    /// let mut map: Hamt<usize, _> = Hamt::new(&store);
+    /// map.set(1, 1).unwrap();
+    /// map.set(4, 2).unwrap();
+    ///
+    /// let mut total = 0;
+    /// map.for_each(&mut |_, v: u64| {
+    ///    total += v;
+    ///    Ok(())
+    /// }).unwrap();
+    /// assert_eq!(total, 3);
+    /// ```
+    #[inline]
+    pub fn for_each<F, V>(&self, f: &mut F) -> Result<(), String>
```

Comment:
couldn't impl the Iterator trait instead?

Reply:
no because all of the elements are not in memory. Also this would be done on top of this, the reason for this is to be able to more closely match the spec.